torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -44,76 +44,94 @@ def _recursive_copy_to_device(value: Any, non_blocking: bool, device: torch.devi
 
 
 def _is_trainable(param: torch.Tensor) -> bool:
+    r"""
+    Returns if a parameter is trainable, where trainability is equivalent to
+    requiring a gradient.
+    """
     return param.requires_grad
 
 
-def _broadcast_object(
-    obj: Any,
-    src_rank: int,
-    group: object = dist.group.WORLD,
-    dist_device: torch.device = torch.device("cpu"),
-) -> Any:
+def _broadcast_object(obj: Any, src_rank: int,
```

Comment:
this is not needed anymore with modern pytorch, it was there initially because this functionality was not present, I think that you could remove it altogether

Reply:
I looked into replacing `_broadcast_object()` with `torch.distributed.broadcast_object_list()` as [suggested](https://github.com/pytorch/pytorch/pull/60285/files/554891f6faa764c76dec4afb1107cb5aa88ef589#r654825551) by @mrshenli. However, `broadcast_object_list()` yields a significant slowdown in my preliminary testing. Precisely, running `test_collect_shards()` on the AI AWS cluster with a world size of 4 regresses from ~6 seconds to ~18 seconds (where the affected function is `consolidate_state_dict()`).

I am unable to resolve the issue, but this is what I found: There are two (possibly-related) points of slowdown in `broadcast_object_list()`:
1) The implicit `torch.cuda.synchronize()` resulting from [moving data from `object_tensor_sizes` from GPU to CPU via `.item()`](https://github.com/pytorch/pytorch/blob/5824a866b72c251ad47a9c16dc652e49cfd7e234/torch/distributed/distributed_c10d.py#L1781) takes ~4 seconds for one of the broadcasts on ranks 1 and 2 (while generally it takes ~1 millisecond). This confuses me since `_broadcast_object()` follows the exact same data movement pattern, yet it does not suffer from a slow synchronization.
2) Unlike `_broadcast_object()`, `broadcast_object_list()` pickles/unpickles the object via [`_object_to_tensor()`](https://github.com/pytorch/pytorch/blob/5824a866b72c251ad47a9c16dc652e49cfd7e234/torch/distributed/distributed_c10d.py#L1757)/[`_tensor_to_object()`](https://github.com/pytorch/pytorch/blob/5824a866b72c251ad47a9c16dc652e49cfd7e234/torch/distributed/distributed_c10d.py#L1797), respectively. `_tensor_to_object()` consistently takes ~4 seconds. 
I am unable to see any connection between the two, but the similarity between their latencies is curious. Together, these two points cause `consolidate_state_dict()` to take >4 seconds when using `broadcast_object_list()`, while it only takes a few milliseconds when using `_broadcast_object()`. 

Any help investigating this is greatly appreciated!

NB: Without pickling, the tensor allocations are 939, 939, and 1131 bytes, while with pickling, the allocations are 571, 591, and 755 bytes.

Reply:
I think I found the issue, and fixing this will require changing code in `distributed_c10d.py`. Here is my hypothesis:

Currently, `broadcast_object_list()` takes care to specially handle the NCCL backend case, making sure to move `object_sizes_tensor` and `object_tensor` to device memory before broadcasting. Upon receiving the broadcasted tensor, the code calls `_tensor_to_object()`, which loads the tensor to an object. The existing implementation uses a `Pickler` and `Unpickler` to save/load, but I have verified that the slowdown behavior is replicated using `torch.save()` and `torch.load()`, which itself[ uses pickling](https://pytorch.org/docs/stable/_modules/torch/serialization.html#load). For the remainder, I will assume a `torch.save()`/`torch.load()`-based implementation of `_object_to_tensor()` and `_tensor_to_object()`.

The crux is the default behavior of `torch.load()` (and presumably `Unpickler.load()`), which deserializes (on CPU) and loads to the device **where the tensor was saved**. Suppose that the world size is 4. Then, in the received `local_state_dict` on rank `j` from the broadcast, the contained tensors are loaded to `cuda:i` where `i != j`. The time bottleneck appears to be exactly in loading these received objects to device, taking `> 3.5` seconds. 

The bandaid fix I have right now is changing:
* `_object_to_tensor(obj)` to `_object_to_tensor(obj, pickle=True)`
* `_tensor_to_object(tensor, tensor_size)` to `_tensor_to_object(tensor, tensor_size, pickle=True, device=None)`

If `pickle == True`, then the behavior is as-is to preserve backward compatibility. Otherwise, `torch.save(obj, f)` and `torch.load(io.BytesIO(buf), map_location=device)` are used instead of the `_pickler` and `_unpickler`, respectively. Then, `broadcast_object_list()` passes in `device=current_device` to `_tensor_to_object()` if using NCCL backend. I have confirmed that this change removes the slowdown.

However, this solution is only a temporary fix, and I think a more thorough examination is needed. I would need to see if a solution is feasible while still using the `Pickler`/`Unpickler` (though at first glance this is not promising since the `Unpickler.load()` provides no options).

I do wonder how much this behavior has affected the speed of other use cases. It seems to be pretty common that the received tensor from the broadcast is from a different device -- in fact, that should be the typical case?

Reply:
Discussed with @andwgu offline on this. Since there are gaps in existing `broadcast_object_list()`, we are going to keep the current `_broadcast_object` method. Later, if bandwidth allows, we can work on a followup PR to improve `broadcast_object_list()` and use it here.